pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstanceassembler/FormInstanceCreator.java

```diff
@@ -0,0 +1,106 @@
+        FormActivityDef formActivityDef = context.getFormActivityDef();
+
+        FormInstance formInstance = new FormInstance(
+                instanceDto.getParticipantId(),
+                instanceDto.getId(),
+                instanceDto.getActivityId(),
+                instanceDto.getActivityCode(),
+                formActivityDef.getFormType(),
+                instanceDto.getGuid(),
+                activityI18nDetail != null ? activityI18nDetail.getTitle() : null,
+                activityI18nDetail != null ? activityI18nDetail.getSubtitle() : null,
+                instanceDto.getStatusType() != null ? instanceDto.getStatusType().name() : null,
+                instanceDto.getReadonly(),
+                formActivityDef.getListStyleHint(),
+                getTemplateId(formActivityDef.getReadonlyHintTemplate()),
+                formActivityDef.getIntroduction() != null ? formActivityDef.getIntroduction().getSectionId() : null,
+                formActivityDef.getClosing() != null ? formActivityDef.getClosing().getSectionId() : null,
+                instanceDto.getCreatedAtMillis(),
+                instanceDto.getFirstCompletedAt(),
+                getTemplateId(formActivityDef.getLastUpdatedTextTemplate()),
+                formActivityDef.getLastUpdated(),
+                formActivityDef.canDeleteInstances(),
+                formActivityDef.isFollowup(),
+                instanceDto.isHidden(),
+                formActivityDef.isExcludeFromDisplay(),
+                instanceDto.getSectionIndex()
+        );
+        return formInstance;
+    }
+
+    private ActivityI18nDetail getActivityI18nDetail() {
+        Optional<ActivityI18nDetail> activityI18nDetail =
+                context.getHandle().attach(ActivityI18nDao.class).findDetailByLanguageCodeIdAndActivityId(
+                        context.getLangCodeId(), context.getActivityInstanceDto().getActivityId());
+        return activityI18nDetail.get();
+    }
+
+    private void addChildren(FormInstance formInstance) {
+        FormActivityDef formActivityDef = context.getFormActivityDef();
+        FormSectionCreator formSectionCreator = new FormSectionCreator(context);
+        formInstance.setIntroduction(formSectionCreator.createSection(formActivityDef.getIntroduction()));
+        formInstance.setClosing(formSectionCreator.createSection(formActivityDef.getClosing()));
+        formActivityDef.getSections().forEach(s -> {
+            formInstance.getBodySections().add(formSectionCreator.createSection(s));
+        });
+    }
+
+    private void updateBlockStatuses(FormInstance formInstance) {
+        formInstance.updateBlockStatuses(
+                context.getHandle(),
+                context.getInterpreter(),
+                context.getUserGuid(),
+                context.getOperatorGuid(),
+                context.getActivityInstanceDto().getGuid(),
+                null);
+    }
+
+    private void render(FormInstance formInstance) {
```

Comment:
@yufengwng in this method rendered `readonlyHint` and `activityDefinitionLastUpdatedText`. Currently the result is the same as during ActivityInstance normal reading (the texts are identical). But yet I'm not sure if it's correct from the versions perspective etc.. I need to check still. And as you see here I just call template.render(0 and then HtmlConverter. So it looks quite similar to formInstance.renderContent() logic but I think I need more deeper understanding of `activityDefinitionLastUpdated` rendering (because it depends on `activityDefinitionLastUpdated'). So, this logic I definitely need to fix in my code still.

Reply:
Take a look at `FormActivityDao::findDefByDtoAndVersion()`. When we build a `FormActivityDef` object, we are already querying for the appropriate version. So the read-only hint and last updated text you get from the `FormActivityDef` object is what you should use.